orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/agent/training_agent.h"
+#include "core/session/IOBinding.h"
+#include "orttraining/training_ops/cpu/controlflow/ort_tasks.h"
+
+namespace onnxruntime {
+namespace training {
+
+TrainingAgent::TrainingAgent(InferenceSession *session) : inference_session_(session) {}
+
+TrainingAgent::~TrainingAgent() {
+  // TODO: Properly cancel outstanding background tasks
+  // Following implementation only handle the case where bg_thread is waiting for backward inputs
+  // Background thread can also be in other states, such as running Forward() or running Backward()
+  std::vector<int64_t> run_ids;
+  {
+    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
+    for (auto it = bg_threads_.begin(); it != bg_threads_.end(); ++it) {
+      run_ids.push_back(it->first);
+    }
+  }
+  for (int64_t run_id : run_ids) {
+    if (!onnxruntime::contrib::OrtTasks::GetInstance().TaskIsCompleted(run_id)) {
+      CancelBackgroundTask(run_id);
+    }
+  }
+};
+
+common::Status TrainingAgent::RunInBackgroundAndWaitForYield(const RunOptions& run_options, onnxruntime::IOBinding& io_binding,
+                                                             std::vector<OrtValue>& user_outputs, int64_t& run_id) {
+  std::promise<void> setup_promise;
+  std::future<void> setup_future = setup_promise.get_future();
+
+  // Passing run_options and io_binding by reference to the bg_thread,
+  // this is ok because they are ORTModule's member, and they are presistent through forward and backward calls
+  auto bg_thread = std::thread([this](std::future<void> setup_future, const RunOptions& run_options, onnxruntime::IOBinding& io_binding) {
+    // wait until task is properly setup
+    setup_future.get();
+
+    common::Status status = inference_session_->Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
```

Comment:
Out of curiosity, could we use `inference_session_->Run(run_options, io_binding);` here to avoid the need for `friend` access?

It looks like that overload of `Run()` does exactly what we need (though perhaps I'm overlooking something?):

https://github.com/microsoft/onnxruntime/blob/9126faa35bdbc5be3add6e7a9da711174f0fc7eb/onnxruntime/core/session/inference_session.cc#L1717-L1718

Reply:
great catch!